crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -177,45 +83,67 @@ pub struct CheckpointExecutorEventLoop {
     /// the final checkpoint will not have been set, thus CheckpointExecutor
     /// will reschedule the last checkpoint and correctly set end_of_epoch.
     end_of_epoch: bool,
+    // If true, need to run crash recovery
+    cold_start: bool,
     metrics: Arc<CheckpointExecutorMetrics>,
 }
 
-impl CheckpointExecutorEventLoop {
-    fn new(
+impl CheckpointExecutor {
+    pub fn new(
         mailbox: broadcast::Receiver<VerifiedCheckpoint>,
-        end_of_epoch_event_sender: broadcast::Sender<Committee>,
         checkpoint_store: Arc<CheckpointStore>,
         authority_state: Arc<AuthorityState>,
         config: CheckpointExecutorConfig,
-        metrics: Arc<CheckpointExecutorMetrics>,
+        prometheus_registry: &Registry,
     ) -> Result<Self, TypedStoreError> {
         Ok(Self {
             mailbox,
-            end_of_epoch_event_sender,
             checkpoint_store,
             authority_state,
             config,
             highest_scheduled_seq_num: None,
             latest_synced_checkpoint: None,
             end_of_epoch: false,
-            metrics,
+            cold_start: true,
+            metrics: CheckpointExecutorMetrics::new(prometheus_registry),
+        })
+    }
+
+    pub fn new_for_tests(
+        mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+    ) -> Result<Self, TypedStoreError> {
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            authority_state,
+            config: Default::default(),
+            highest_scheduled_seq_num: None,
+            latest_synced_checkpoint: None,
+            end_of_epoch: false,
+            cold_start: true,
+            metrics: CheckpointExecutorMetrics::new_for_tests(),
         })
     }
 
-    pub async fn run(mut self) {
-        self.handle_crash_recovery().await.unwrap();
+    pub async fn run_epoch(&mut self) -> Committee {
+        if self.cold_start {
+            let end_of_epoch_recovery = self.handle_crash_recovery().await.unwrap();
+            self.cold_start = false;
 
-        while let Some((last_checkpoint, next_committee)) =
-            self.execute_checkpoints_for_epoch().await
-        {
-            self.reconfig(next_committee, last_checkpoint.epoch()).await;
-            self.end_of_epoch = false;
+            if let Some(committee) = end_of_epoch_recovery {
+                return committee;
+            }
         }
-        // Channel closed
+
+        self.end_of_epoch = false;
+        self.run_epoch_impl().await
     }
 
-    pub async fn handle_crash_recovery(&self) -> SuiResult {
+    async fn handle_crash_recovery(&self) -> SuiResult<Option<Committee>> {
         let local_epoch = self.authority_state.epoch();
```

Comment:
Instead of requesting epoch from authority_state, now that the executor is per-epoch, we should pass in the epoch store and make it a field of the executor. All references to the epoch store could use it.

Reply:
True. Will do